### ISSUE TITLE
pipelines of transformations discussion

### ISSUE THREAD

**Issue by tbreloff** [JuliaML/Losses.jl#12]
Continuing the discussion from #8...

I'd like to plan out the long term goals/terminology of the computation pipeline for fitting and using various ML models.  Here's my current thinking... lets try to poke holes in its viability using some real-world models that should be able to handle:

- Data preprocessing: filtering, projections, NA replacement, column transforms
- Regression variants
- SVM variants
- Random Forest variants
- ANN, RNN variants
- Ensembles, model aggregation

Empirical risk minimization is a (IMO confusing) name for minimizing a function of model loss (we currently call this `PredictionLoss`... do you like `ModelLoss` instead?) and parameter loss: `L(loss, ploss)`.  I would prefer we migrate away from the ERM name, even if it is a "descriptively correct" identifier.  Maybe we could just call it `TotalLoss`:

```
immutable TotalLoss{ML <: ModelLoss, PL <: ParameterLoss} <: Loss
    mloss::ML
    ploss::PL
end
```

### Pipelines of transformations

To sum up a typical real-world problem... we want to build a 2-class classifier for a dataset.  To build a logistic regression:
```
data --> filter_columns --> filter_rows --> replace_NAs --> add_transformed_vars
     --> learnable_affine_transformation --> SigmoidActivation --> LogitDistLoss
target --> LogitDistLoss
```

Now we want to do the same thing for an SVM... it's pretty similar, but there's probably no need to add transformed vars (cross multiplying xi * xj for example) because that's what the kernel is for:
```
data --> filter_columns --> filter_rows --> replace_NAs
     --> learnable_affine_transformation --> LogitMarginLoss
target --> LogitMarginLoss
```

An ANN would be similar to the logistic regression, but there would be many `affine_transformation --> activation` in sequence.

A decision tree may not need the filter/transform steps (or they may be slightly different), but their learnable transformation would be a "tree of filter transformations".

In all of these examples, there are some parts of the pipeline that can be reused.  If one was to create many different models, whether as part of an ensemble or something else, I think treating these pipelines as directed graphs could give some great opportunities for optimization/caching and visualization.

### Ensembles

Now a random forest would be an aggregation of decision trees, so we should be able to cover that as an `EnsembleTransformation`, which takes multiple other transformations as input and outputs a single vector.  In theory, there's no reason we can't add an ensemble as part of a fitting/backprop algorithm, passing error sensitivity/responsibility to individual models through the aggregation.  It would also open up SGD-style training of layers of ensembles (I'm picturing the US Electoral College voting process as an example).

The requirement is that an ensemble shares some basic verbs with "constant transforms" (such as taking a square root) and more complex transforms (like a decision tree).  It must be able to go `forward` to update its internal state and produce an output value, and it must be able to go `backward` to update gradients/etc, compute error responsibilities, and adjust learnable parameters.

### Summary

Anything that takes in an input vector(s) and produces an output vector can be considered a `Transformation`.  If we can be smart about how to connect various transformations during the `forward` and `backward` passes of a learning algorithm, then it opens up some really nice opportunities in building, optimizing, and visualizing modelling pipelines, as well as potentially making ensemble-type aggregation part of the supervised learning step.

**Comment by Evizero** [comment]
I will review this tomorrow. I've been busy with work all day

**Comment by tbreloff** [comment]
Just in case you didn't get an email... I started a gitter room for LearnBase here: https://gitter.im/Evizero/LearnBase.jl.  @joshday and I had a quick conversation, and thought that `Mapping` was probably better terminology as opposed to `Activation`.  Also, I had the thought that maybe there should be separate notation for those cases where there's multiple inputs (for example, elementwise multiplication, which is an integral part of what I'm working on in OnlineAI)... maybe `Aggregation`?

So the type tree might look like:

```
Transformation (abstract)
    Mapping (abstract)
        Identity
        Sigmoid
        Softmax
        ...
        Center
        Normalize
        Sqrt
        ...
        Filter (abstract)
            Dropout
            TreeSplit
            ...
        Kernel (abstract)
            RBF
            Polynomial
            ...
    Aggregation (abstract)
        Concat
        Multiply
        Average
        Vote
        ...
    Learnable (abstract)
        Affine
        Translation
        Scale
        Linear
        ...
```

Each `Transformation` takes input(s), produces output, and has a `forward` and `backward` pass.  Mappings take one input vector and have deterministic output.  Aggregations take many input vectors and have deterministic output.  Learnables take one input vector and have varying output depending on the parameters that have been learned.

I **think** we can represent all the major (non-bayesian?) ML approaches in this way.  Can you think of weirdness that might arise?

**Comment by Evizero** [comment]
One important thing we should incorporate is that we shouldn't expect this package to become the "Base Package for everything". For example, it's not reasonable to assume that JuliaStats and JuliaOpt etc follow our guideline.

What that means is we should find a way to work around that constraint. For example for Kernels I would like to use [MLKernels.jl](https://github.com/trthatcher/MLKernels.jl), which is a really nice package. I don't necessarily think there is a real need for that package to depend on LearnBase

I would like to concentrate on those things that we three actually have a use-case for, and those actually also heavily overlap. I would say that these "things" can more or less be listed under the term "coefficient-based prediction model". These include neural networks, empirical risk minimization, and SVMs. In contrast, I don't have enough practice with trees and the likes to make good design decisions there. I would like this package to grow organically. That doesn't mean that I want to discourage planning ahead, but we shouldn't get lost in the details there.

**Comment by Evizero** [comment]
> do you like ModelLoss instead

I would say `PredictionLoss` is a bit more descriptive than `ModelLoss`, but if you both prefer the later then I'm ok with it.

> I would prefer we migrate away from the ERM name

I'm ok with that. What I need is an object that based on my prediction function, loss, and penalty, conveniently gives me to gradient and the cost for my current parameters and (sub)data (like [here](https://github.com/Evizero/KSVM.jl/blob/master/src/linear/solver/pegasos.jl#L110)). Anyone should be able to use that structure. I do not want to create typical black box package that just offers a high level API. That would be a waste of the language. I still feel strongly that the building blocks should make sense to a machine learning person reading the code.

> Pipelines of transformations

I really like that. But let us try to make this functionality a high-level convenience interface and try to make sure the smaller building blocks like Losses make sense on their own (for example someone wants to implement a simple gradient descent algorithm as an exercise and use the loss functions to see how they behave).

**Comment by Evizero** [comment]
> TotalLoss

What would be the purpose of that structure? Without the prediction model it is no replacement for EmpiricalRisk, and I don't see any other use of it either.

**Comment by Evizero** [comment]
I'm a bit concerned about `Mapping` and `Transformation`. They are not very descriptive to me. Also side by side they seem like the same thing. A mapping is to my knowledge just another word for a function, which can be any input-output transformation outside ML as well.

That said, I don't have a good suggestion either. What about something along the lines of `StaticTransformation` instead of `Mapping` to contrast it to `LearnableTransformation` ?

**Comment by Evizero** [comment]
All in all I like what you are proposing very much. I just want to urge us to make incremental steps and to keep reflecting along the way.

**Comment by joshday** [comment]
I do like `StaticTransformation` / `LearnableTransformation`.

**Comment by tbreloff** [comment]
I can get behind static/learnable transformations. 

I'm not planning on implementing most of this stuff on first pass, but I do think it's important to have a plan for the eventual design (especially if anyone else wants to contribute). I'm pretty close to the organization and functionality that I want in the short term, so I'll probably be switching my focus soon, and coming back when I have time.

**Comment by tbreloff** [comment]
> A mapping is to my knowledge just another word for a function, which can be any input-output transformation outside ML as well.

The difference between "any function" and this type of mapping is that we expect to have a forward and backward pass that makes sense.  The function `f(x) = "hello world"` doesn't make any sense.  So I agree it's a pretty general term, but then again it encompasses a lot of concepts in the type tree I laid out so its ok to be general.

> What about something along the lines of StaticTransformation

I like the idea behind `StaticTransformation`, but I just went to rename the concrete subtypes, and I didn't like any of `SigmoidStatic` or `SigmoidStaticTransformation`.  I suppose it could be `SigmoidTransformation <: StaticTransformation`, but maybe it makes more sense as `SigmoidMapping <: Mapping`.  Thoughts?

**Comment by tbreloff** [comment]
> What that means is we should find a way to work around that constraint. For example for Kernels I would like to use MLKernels.jl, which is a really nice package. I don't necessarily think there is a real need for that package to depend on LearnBase

Yeah I hear that.  In this case (if we decide we want `Kernel <: Transformation`), then I think we could bite the bullet and just wrap the MLKernels object with a LearnBase object (possibly doing it within a `@require MLKernels` block)? Though I don't know enough about MLKernels to say how easy that would be.

**Comment by Evizero** [comment]
> I'm not planning on implementing most of this stuff on first pass, but I do think it's important to have a plan for the eventual design (especially if anyone else wants to contribute)

:+1:

I will then refactor KSVM accordingly, provide feedback, and then probably do the next iteration

> I didn't like any of SigmoidStatic or SigmoidStaticTransformation

What about `SigmoidActivation <: Activation <: StaticTransformation` ? It does make sense to group the activations functions somehow

**Comment by Evizero** [comment]
Concerning Kernels, I am not yet sure how to best incorporate them. As it is now I don't see a reason for LearnBase to know about them at all. I will know more once I refactored to the new changes.

**Comment by tbreloff** [comment]
> What I need is an object that based on my prediction function, loss, and penalty, conveniently gives me the gradient and the cost for my current parameters and (sub)data (like here). 

> > TotalLoss

> What would be the purpose of that structure?

If I understand, you want to be able compute the gradient:

```
L(x,w) = mloss(f(x,w),w) + ploss(w)
∇ᵢ = ∂L / ∂wᵢ
```

Using an ERM concept, you would have an object `erm` that stores `f`, `mloss`, and `ploss`... you would then pass it into a `grad` function along with the input and weights: `grad(erm, x, w)`.  This means you both have to re-compute the function `f` in order to compute the gradient, and you also need to have your full set of learnable parameters up front.

If, however, the core interface was `grad(totloss, y)`, then you have a lot more flexibility, and you could always include the convenience `grad(erm, x, w) = grad(totalloss(erm), predict(erm, x, w))` for super-simple prediction models.  For my purposes, it's rare that I would be able to use that convenience, as the `predict` happens in multiple stages, and the `grad` happens in many places with many parameter matrices.

So I think `TotalCost` could be defined as above (containing a `ModelLoss` and `ParameterLoss`), and an ERM could contain a `TotalLoss` and `LearnableTransformation` (for those cases when you only have a single transformation to learn).  Does this make sense to you?

**Comment by Evizero** [comment]
> This means you both have to re-compute the function f in order to compute the gradient,

I don't have to do that, since I can pass the predictions along. (see [here](https://github.com/Evizero/KSVM.jl/blob/master/src/linear/solver/pegasos.jl#L106-L110))

> and you also need to have your full set of learnable parameters up front

Not sure what you mean there. All `EmpiricalRisk` does is specify the loss, model, and penalty (see [here](https://github.com/Evizero/KSVM.jl/blob/master/src/linear/solver/pegasos.jl#L65-L68), but I do admit this is not ideal code as I simply wanted to test if it works as intended). It doesn't care about how `theta` looks like, as long as the parameters are compatible to each other.

> TotalLoss

The main reason why I am reserved about total loss is that I feel that the prediction model is much closer to the model loss than the model loss to the parameter loss. Just think about the convenient gradient of crossentropy plus sigmoid.

> `grad(totloss, y)`

Two things I don't understand

1. How do you evaluate the `grad` for the `ParameterLoss` without `w`? Or did you mean to write `grad(totloss, y, w)` ?

2. How is this supposed to work without knowing what the prediction model is? One needs the chain rule to really evaluate the gradient for the parameters given the model loss and prediction model. For a linear model it might always be `X * deriv(mloss, output, target)` but that's a pretty strong assumption to enforce. Also,again thinking about the convenient gradient of crossentropy plus sigmoid, one wants to make use of these analytics shortcuts.

> So I think TotalCost could be defined as above (containing a ModelLoss and ParameterLoss), and an ERM could contain a TotalLoss and LearnableTransformation

I still don't see the upside, which may just be my predispositions. Maybe if you tell me what the benefit would be to you I could understand it better

**Comment by tbreloff** [comment]
I don't have time for a full response, but I think you're right... ParameterLoss and the associated LearnableTransformation are tightly coupled... Not ploss and mloss. So totalloss doesn't help much. 

I still think that ERM is only useful in certain single-transformation models though... Better to keep these components separate in most cases I think.

**Comment by Evizero** [comment]
> I still think that ERM is only useful in certain single-transformation models though... Better to keep these components separate in most cases I think

no disagreement there. but something like the current `EmpiricalRisk` should still exist as a choice. Like a convenience layer on top of the losses and transformation

**Comment by Evizero** [comment]
delegated to other packges
